### PR TITLE
DSND-2375: Retry 404 errors in company links consumer

### PR DESCRIPTION
**Pull request by crowlandsCH**
* Added retry and fixed some Sonar issues.
* NOT locally tested yet.

[DSND-2375](https://companieshouse.atlassian.net/browse/DSND-2375)

[DSND-2375]: https://companieshouse.atlassian.net/browse/DSND-2375?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/company/links/service/AddExemptionsClient.java [hunk 1]
```diff
@@ -53,6 +53,10 @@ public void patchLink(PatchLinkRequest linkRequest) {
             } else if (ex.getStatusCode() == 404) {
                 logger.info("HTTP 404 Not Found returned; "
                         + "company profile does not exist", DataMapHolder.getLogMap());
+                throw new RetryableErrorException(
+                        String.format("Company profile [%s] does not exist"
+                                        + " when processing add filing history link request",
```
[the remark this answers is not in the document]
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/AddExemptionsClient.java:58]: done

File: src/main/java/uk/gov/companieshouse/company/links/service/AddOfficersClient.java [hunk 2]
```diff
@@ -52,6 +52,10 @@ public void patchLink(PatchLinkRequest linkRequest) {
             } else if (ex.getStatusCode() == 404) {
                 logger.info("HTTP 404 Not Found returned; "
                         + "company profile does not exist", DataMapHolder.getLogMap());
+                throw new RetryableErrorException(
+                        String.format("Company profile [%s] does not exist"
+                                        + " when processing add filing history link request",
```
[the remark this answers is not in the document]
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/AddOfficersClient.java:57]: done

File: src/main/java/uk/gov/companieshouse/company/links/service/AddPscClient.java [hunk 3]
```diff
@@ -50,6 +50,10 @@ public void patchLink(PatchLinkRequest linkRequest) {
                         + "company profile already has a PSC link");
             } else if (ex.getStatusCode() == 404) {
                 logger.info("HTTP 404 Not Found returned; company profile does not exist");
+                throw new RetryableErrorException(
+                        String.format("Company profile [%s] does not exist"
+                                        + " when processing add filing history link request",
```
[the remark this answers is not in the document]
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/AddPscClient.java:55]: done

File: src/main/java/uk/gov/companieshouse/company/links/service/AddStatementsClient.java [hunk 4]
```diff
@@ -52,6 +52,10 @@ public void patchLink(PatchLinkRequest linkRequest) {
             } else if (ex.getStatusCode() == 404) {
                 logger.info("HTTP 404 Not Found returned; company profile does not exist",
                         DataMapHolder.getLogMap());
+                throw new RetryableErrorException(
+                        String.format("Company profile [%s] does not exist"
+                                        + " when processing add filing history link request",
```
[the remark this answers is not in the document]
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/AddStatementsClient.java:57]: done

File: src/main/java/uk/gov/companieshouse/company/links/service/PscListClient.java [hunk 5]
```diff
@@ -58,7 +58,7 @@ public PscList getPscs(PatchLinkRequest linkRequest) {
                         && !ex.getContent().contains("company-psc-not-found")) {
                     if ((ex.getHeaders().containsKey(HttpHeaders.CONTENT_LENGTH)
                             && ex.getHeaders().getContentLength() > 0)
-                            || (ex.getContent() != null && ex.getContent().length() > 0)) {
+                            || (ex.getContent() != null && !ex.getContent().isEmpty())) {
```
[the remark this answers is not in the document]
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/PscListClient.java:61]: done

File: src/main/java/uk/gov/companieshouse/company/links/service/PscListClient.java [hunk 6]
```diff
@@ -58,7 +58,7 @@ public PscList getPscs(PatchLinkRequest linkRequest) {
                         && !ex.getContent().contains("company-psc-not-found")) {
                     if ((ex.getHeaders().containsKey(HttpHeaders.CONTENT_LENGTH)
```
[the remark this answers is not in the document]
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/PscListClient.java:59]: done

File: src/main/java/uk/gov/companieshouse/company/links/service/AddExemptionsClient.java [hunk 7]
```diff
@@ -53,6 +53,10 @@ public void patchLink(PatchLinkRequest linkRequest) {
             } else if (ex.getStatusCode() == 404) {
                 logger.info("HTTP 404 Not Found returned; "
                         + "company profile does not exist", DataMapHolder.getLogMap());
+                throw new RetryableErrorException(
```
Comment (johncook-ch) [inline src/main/java/uk/gov/companieshouse/company/links/service/AddExemptionsClient.java:56]: SonarQube is flagging these as duplicate code and failing. A solution could be to add another constructor that takes arguments that builds the message. Something like this: 
```
public RetryableErrorException(String companyNumber, String linkType, Exception exception) {
    super(String.format("Company profile [%s] does not exist when processing add %s link request",
            companyNumber, linkType), exception);
}
```

File: src/main/java/uk/gov/companieshouse/company/links/service/AddOfficersClient.java [hunk 8]
```diff
@@ -52,6 +52,10 @@ public void patchLink(PatchLinkRequest linkRequest) {
             } else if (ex.getStatusCode() == 404) {
                 logger.info("HTTP 404 Not Found returned; "
                         + "company profile does not exist", DataMapHolder.getLogMap());
+                throw new RetryableErrorException(
```
Comment (johncook-ch) [inline src/main/java/uk/gov/companieshouse/company/links/service/AddOfficersClient.java:55]: See comment on AddExemptionsClient.java

File: src/main/java/uk/gov/companieshouse/company/links/service/AddPscClient.java [hunk 9]
```diff
@@ -50,6 +50,10 @@ public void patchLink(PatchLinkRequest linkRequest) {
                         + "company profile already has a PSC link");
             } else if (ex.getStatusCode() == 404) {
                 logger.info("HTTP 404 Not Found returned; company profile does not exist");
+                throw new RetryableErrorException(
```
Comment (johncook-ch) [inline src/main/java/uk/gov/companieshouse/company/links/service/AddPscClient.java:53]: See comment on AddExemptionsClient.java

File: src/main/java/uk/gov/companieshouse/company/links/service/AddStatementsClient.java [hunk 10]
```diff
@@ -52,6 +52,10 @@ public void patchLink(PatchLinkRequest linkRequest) {
             } else if (ex.getStatusCode() == 404) {
                 logger.info("HTTP 404 Not Found returned; company profile does not exist",
                         DataMapHolder.getLogMap());
+                throw new RetryableErrorException(
```
Comment (johncook-ch) [inline src/main/java/uk/gov/companieshouse/company/links/service/AddStatementsClient.java:55]: See comment on AddExemptionsClient.java

**Review by johncook-ch — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/companieshouse/company/links/service/AddExemptionsClient.java (hunk 7 above)
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/AddExemptionsClient.java:56]: done

File: src/main/java/uk/gov/companieshouse/company/links/service/AddOfficersClient.java (hunk 8 above)
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/AddOfficersClient.java]: done

File: src/main/java/uk/gov/companieshouse/company/links/service/AddStatementsClient.java (hunk 10 above)
Comment (crowlandsCH) [reply src/main/java/uk/gov/companieshouse/company/links/service/AddStatementsClient.java]: done

**Review by johncook-ch — APPROVED** [review]

File: src/test/java/uk/gov/companieshouse/company/links/service/AddExemptionsClientTest.java [hunk 11]
```diff
@@ -55,6 +55,9 @@ class AddExemptionsClientTest {
     @Mock
     private Logger logger;
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/company/links/service/AddExemptionsClientTest.java:56]: Unused mock

File: src/test/java/uk/gov/companieshouse/company/links/service/AddStatementsClientTest.java [hunk 12]
```diff
@@ -72,20 +68,21 @@ class AddStatementsClientTest {
     @Mock
     private StatementList getData;
 
-    private List<Statement> statementsList;
-
     @Mock
     private HttpClient httpClient;
 
     @Mock
```
Comment (cgregoryCH) [inline src/test/java/uk/gov/companieshouse/company/links/service/AddStatementsClientTest.java:74]: Same here

**Review by cgregoryCH — CHANGES_REQUESTED** [review]

**Review by cgregoryCH — APPROVED** [review]

**Review by johncook-ch — APPROVED** [review]

**Pull request merged**